File: Sources/Helmsman/Catalog/NodeFit.swift
import Foundation

/// Capacity headroom for one node, in app-relative terms.
struct NodeFit: Hashable, Identifiable {
    let node: Node
    /// Free CPU after subtracting pending+running pod requests.
    let freeCPU: Double
    /// Free memory in bytes after subtracting pending+running pod requests.
    let freeMemoryBytes: Double
    /// Total allocatable CPU (cores) on the node.
    let allocatableCPU: Double
    /// Total allocatable memory (bytes) on the node.
    let allocatableMemoryBytes: Double
    /// Free disk in bytes (actual available from the kubelet Summary API when
    /// known, else allocatable ephemeral-storage).
    let freeDiskBytes: Double
    /// Total disk in bytes (filesystem capacity when known, else allocatable
    /// ephemeral-storage). 0 when neither is available.
    let allocatableDiskBytes: Double
    /// Actual used disk in bytes from the Summary API; nil when usage is unknown
    /// (only then is disk left out of headroom + fit gating).
    let usedDiskBytes: Double?
    /// True = the app's CPU, memory AND (when known) disk requests fit here.
    let canHost: Bool
    /// True = `NoSchedule` (or `NoExecute`) taint excludes this node.
    let tainted: Bool
    /// True = node spec marks it unschedulable.
    let cordoned: Bool

    var id: String { node.id }

    /// True when the app can be pinned here: it fits and the node isn't
    /// tainted or cordoned. Same predicate used to rank nodes and to gate
    /// node-pin selection in the catalog UI.
    var eligible: Bool { canHost && !tainted && !cordoned }

    /// 0..1 — average fraction of headroom remaining across the resources we can
    /// measure (CPU, memory, and disk when actual usage is known). Used purely to
    /// sort nodes that all qualify. Disk only joins the average when
    /// `usedDiskBytes` is known, so nodes without Summary-API data rank exactly as
    /// before (CPU+memory only).
    var headroomScore: Double {
        var ratios: [Double] = []
        if allocatableCPU > 0 { ratios.append(max(0, freeCPU / allocatableCPU)) }
        if allocatableMemoryBytes > 0 { ratios.append(max(0, freeMemoryBytes / allocatableMemoryBytes)) }
        if usedDiskBytes != nil, allocatableDiskBytes > 0 { ratios.append(max(0, freeDiskBytes / allocatableDiskBytes)) }
        return ratios.isEmpty ? 0 : ratios.reduce(0, +) / Double(ratios.count)
    }
}

/// Outcome of fitting one `CatalogApp` against current cluster state.
struct FitResult: Hashable {
    /// One entry per node, ordered so eligible (`canHost && !tainted && !cordoned`)
    /// nodes come first, sorted by `headroomScore` descending. Remaining
    /// nodes follow in arbitrary stable order.
    let perNode: [NodeFit]
    /// First eligible node, or nil when nothing fits.
    let recommended: NodeFit?

    var anyFits: Bool { recommended != nil }

    /// Cluster-wide categorical fit indicator for the catalog card dot.
    /// .green = at least one eligible node has >50% headroom remaining.
    /// .yellow = at least one eligible node fits but headroom is tight.
    /// .red = no eligible node fits.
    var dot: FitDot {
        guard let rec = recommended else { return .red }
        return rec.headroomScore >= 0.5 ? .green : .yellow
    }
}

enum FitDot: Hashable {
    case green, yellow, red
}

/// Compute per-node fit for an app against the live cluster state.
/// Pure — no side effects on the cache. Inputs are the app's requirements
/// plus snapshots of `nodes` and `pods` from `ClusterCache`.
func nodeFit(app: CatalogApp, nodes: [Node], pods: [Pod], nodeDisk: [String: NodeDiskUsage] = [:]) -> FitResult {
    let appCPU = ResourceQuantity.cpuCores(app.requirements.cpuRequest)
    let appMem = ResourceQuantity.bytes(app.requirements.memoryRequest)
    // The app's primary PVC ask, in bytes (0 when the app declares no storage).
    let appDisk = Double(app.requirements.storageGiB ?? 0) * 1024 * 1024 * 1024

    // Pre-aggregate pod requests by node name. Skip terminal pods so a node
    // with a recently-completed Job doesn't look saturated.
    var cpuUsedByNode: [String: Double] = [:]
    var memUsedByNode: [String: Double] = [:]
    for pod in pods {
        guard let nodeName = pod.spec?.nodeName else { continue }
        let phase = pod.status?.phase ?? ""
        if phase == "Succeeded" || phase == "Failed" { continue }
        for container in pod.spec?.containers ?? [] {
            if let cpu = container.resources?.requests?["cpu"] {
                cpuUsedByNode[nodeName, default: 0] += ResourceQuantity.cpuCores(cpu)
            }
            if let mem = container.resources?.requests?["memory"] {
                memUsedByNode[nodeName, default: 0] += ResourceQuantity.bytes(mem)
            }
        }
    }

    let fits: [NodeFit] = nodes.map { node in
        let name = node.metadata.name
        let allocCPU = ResourceQuantity.cpuCores(node.status?.allocatable?["cpu"] ?? "0")
        let allocMem = ResourceQuantity.bytes(node.status?.allocatable?["memory"] ?? "0")
        let freeCPU = max(0, allocCPU - (cpuUsedByNode[name] ?? 0))
        let freeMem = max(0, allocMem - (memUsedByNode[name] ?? 0))
        // Disk: prefer real Summary-API usage; fall back to allocatable
        // ephemeral-storage (capacity only, no usage) so display still works.
        let disk = nodeDisk[name]
        let allocDisk = disk?.capacityBytes
            ?? ResourceQuantity.bytes(node.status?.allocatable?["ephemeral-storage"] ?? "0")
        let freeDisk = disk?.availableBytes ?? allocDisk
        // Only gate on disk when we have data AND the app asks for storage; an
        // unknown filesystem or networked PVC must never falsely exclude a node.
        let diskFits = allocDisk <= 0 || appDisk <= 0 || freeDisk >= appDisk
        let tainted = (node.spec?.taints ?? []).contains { $0.effect == "NoSchedule" || $0.effect == "NoExecute" }
        let cordoned = node.spec?.unschedulable == true
        let canHost = freeCPU >= appCPU && freeMem >= appMem && diskFits && node.isReady
        return NodeFit(
            node: node,
            freeCPU: freeCPU,
            freeMemoryBytes: freeMem,
            allocatableCPU: allocCPU,
            allocatableMemoryBytes: allocMem,
            freeDiskBytes: freeDisk,
            allocatableDiskBytes: allocDisk,
            usedDiskBytes: disk?.usedBytes,
            canHost: canHost,
            tainted: tainted,
            cordoned: cordoned
        )
    }

    // Eligible = can host AND not tainted AND not cordoned. Sort eligible
    // first by headroom desc, then everything else trailing in name order
    // so the per-node list is deterministic for tests + UI.
    let eligible = fits
        .filter { $0.eligible }
        .sorted { $0.headroomScore > $1.headroomScore }
    let ineligible = fits
        .filter { !$0.eligible }
        .sorted { $0.node.metadata.name < $1.node.metadata.name }

    return FitResult(perNode: eligible + ineligible, recommended: eligible.first)
}

File: apps/web/src/panels/catalog/nodeFit.ts
/**
 * NodeFit — capacity-fit logic for catalog apps.
 *
 * Pure TypeScript port of `Sources/Rigel/Catalog/NodeFit.swift`.
 * No side effects. Inputs: a CatalogApp + snapshots of nodes and pods.
 *
 * Disk: the web has no kubelet Summary API, so `usedDiskBytes` is always
 * undefined and disk headroom is omitted from `headroomScore`. `freeDiskBytes`
 * falls back to `allocatableDiskBytes` (ephemeral-storage from allocatable).
 * This matches Swift's fallback path exactly.
 */

import { parseCpuCores, parseBytes, isReady } from "@/panels/nodes/nodeDisplay";
import type { Node } from "@/panels/nodes/types";
import type { Pod } from "@/panels/pods/types";
import type { CatalogApp } from "@rigel/catalog";

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export interface NodeFitEntry {
  node: Node;
  freeCPU: number;
  freeMemoryBytes: number;
  allocatableCPU: number;
  allocatableMemoryBytes: number;
  freeDiskBytes: number;
  allocatableDiskBytes: number;
  /** Always undefined on the web — no kubelet Summary API. */
  usedDiskBytes?: number;
  canHost: boolean;
  tainted: boolean;
  cordoned: boolean;
  /** canHost && !tainted && !cordoned */
  eligible: boolean;
  /**
   * 0–1 average headroom fraction across measurable resources (CPU + memory;
   * disk only when usedDiskBytes is known — it won't be on the web). Used to
   * sort eligible nodes. Clamped ≥ 0.
   */
  headroomScore: number;
}

export type FitDot = "green" | "yellow" | "red";

export interface FitResult {
  /**
   * One entry per node. Eligible nodes first, sorted by headroomScore desc.
   * Ineligible nodes follow, sorted by node name asc.
   */
  perNode: NodeFitEntry[];
  /** First eligible node, or null when nothing fits. */
  recommended: NodeFitEntry | null;
  /** True when at least one eligible node was found. */
  anyFits: boolean;
  /**
   * Cluster-wide categorical fit indicator for the catalog card dot.
   * green  = recommended.headroomScore >= 0.5
   * yellow = recommended exists but headroomScore < 0.5
   * red    = no recommended node
   */
  dot: FitDot;
}

// ---------------------------------------------------------------------------
// Implementation
// ---------------------------------------------------------------------------

/**
 * Compute per-node fit for a catalog app against current cluster state.
 * Pure — no network calls, no side effects.
 */
export function nodeFit(app: CatalogApp, nodes: Node[], pods: Pod[]): FitResult {
  const appCPU = parseCpuCores(app.requirements.cpuRequest);
  const appMem = parseBytes(app.requirements.memoryRequest);
  // 0 when the app declares no storage (stateless).
  const appDisk =
    app.requirements.storageGiB != null && app.requirements.storageGiB > 0
      ? app.requirements.storageGiB * 1024 * 1024 * 1024
      : 0;

  // Pre-aggregate pod resource requests by node name.
  // Skip terminal pods (Succeeded / Failed) so a completed Job doesn't make
  // a node look saturated — mirrors Swift exactly.
  const cpuUsedByNode: Record<string, number> = {};
  const memUsedByNode: Record<string, number> = {};

  for (const pod of pods) {
    const nodeName = pod.spec?.nodeName;
    if (!nodeName) continue;
    const phase = pod.status?.phase ?? "";
    if (phase === "Succeeded" || phase === "Failed") continue;
    for (const container of pod.spec?.containers ?? []) {
      const cpuReq = container.resources?.requests?.["cpu"];
      const memReq = container.resources?.requests?.["memory"];
      if (cpuReq) cpuUsedByNode[nodeName] = (cpuUsedByNode[nodeName] ?? 0) + parseCpuCores(cpuReq);
      if (memReq) memUsedByNode[nodeName] = (memUsedByNode[nodeName] ?? 0) + parseBytes(memReq);
    }
  }

  const fits: NodeFitEntry[] = nodes.map((node) => {
    const name = node.metadata.name;
    const allocCPU = parseCpuCores(node.status?.allocatable?.["cpu"]);
    const allocMem = parseBytes(node.status?.allocatable?.["memory"]);
    const freeCPU = Math.max(0, allocCPU - (cpuUsedByNode[name] ?? 0));
    const freeMem = Math.max(0, allocMem - (memUsedByNode[name] ?? 0));

    // Disk: no Summary API on web → freeDisk = allocDisk (fallback path)
    const allocDisk = parseBytes(node.status?.allocatable?.["ephemeral-storage"]);
    const freeDisk = allocDisk; // usedDiskBytes always undefined on web

    // diskFits: skip gating when node has no ephemeral-storage declared
    // (allocDisk=0) or when the app requests no storage (appDisk=0).
    const diskFits = allocDisk <= 0 || appDisk <= 0 || freeDisk >= appDisk;

    // tainted = any taint with effect NoSchedule or NoExecute
    const tainted = (node.spec?.taints ?? []).some(
      (t) => t.effect === "NoSchedule" || t.effect === "NoExecute",
    );
    const cordoned = node.spec?.unschedulable === true;

    const canHost = freeCPU >= appCPU && freeMem >= appMem && diskFits && isReady(node);

    const eligible = canHost && !tainted && !cordoned;

    // headroomScore: average of measurable ratios, clamped ≥ 0
    const ratios: number[] = [];
    if (allocCPU > 0) ratios.push(Math.max(0, freeCPU / allocCPU));
    if (allocMem > 0) ratios.push(Math.max(0, freeMem / allocMem));
    // disk skipped — usedDiskBytes is always undefined on the web
    const headroomScore = ratios.length === 0 ? 0 : ratios.reduce((a, b) => a + b, 0) / ratios.length;

    return {
      node,
      freeCPU,
      freeMemoryBytes: freeMem,
      allocatableCPU: allocCPU,
      allocatableMemoryBytes: allocMem,
      freeDiskBytes: freeDisk,
      allocatableDiskBytes: allocDisk,
      usedDiskBytes: undefined,
      canHost,
      tainted,
      cordoned,
      eligible,
      headroomScore,
    };
  });

  // Sort: eligible first by headroomScore desc; ineligible after by node name asc
  const eligibleEntries = fits
    .filter((e) => e.eligible)
    .sort((a, b) => b.headroomScore - a.headroomScore);
  const ineligibleEntries = fits
    .filter((e) => !e.eligible)
    .sort((a, b) => a.node.metadata.name.localeCompare(b.node.metadata.name));

  const perNode = [...eligibleEntries, ...ineligibleEntries];
  const recommended = eligibleEntries[0] ?? null;

  const dot: FitDot = recommended == null ? "red" : recommended.headroomScore >= 0.5 ? "green" : "yellow";

  return { perNode, recommended, anyFits: recommended !== null, dot };
}
